conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/CollectionValidationModule.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.serialization;
+
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Validates that collections are deserialized into data that Conjure allows. This module provides <i>soft</i>
+ * validation, logging instead of breaking functionality. Over time this will transition to strict validation,
+ * throwing exceptions when incorrect data is received.
+ * This implementation allows us to discover code which leverages incorrect behavior and remediate without
+ * breaking functionality.
+ *
+ * What is validated?
+ * <ul>
+ *     <li>Collection values must not be null</li>
+ *     <li>TODO: Values are not mutated (after deserialization)</li>
```

Comment:
Open question: Do we want hard, or soft validation on mutability? Unlike null values, this is related entirely to implementation, not wire format.